### ISSUE TITLE
Preface generated class and interface declarations with @:rtti

### ISSUE THREAD

**Issue by elliott5** [go2hx/go2hx#38]
I'm planning to fork this repo and write the prototype `reflect` module in the correct location.

It would be really helpful to my testing @PXshadow  if you could preface generated `class` and `interface` code with `@:rtti`.

There will be more code generation tweaks required, if the prototype approach works.

**Comment by PXshadow** [comment]
@elliott5 Fantastic! Let me know if you need any more changes on my side to help you with the prototype.

**Comment by elliott5** [comment]
Unfortunately `@:rtti` generation is not working yet on my ...__extension classes, don't know why.

Also, https://github.com/go2hx/go2hx/issues/42#issuecomment-826342021 contains details of how the non-struct named types / ___extension classes should be generated to work with the prototype reflect package/module.

**Comment by PXshadow** [comment]
@elliott5 not working yet in what capacity? I have tested it and ``@:rtti`` seems to be added to _extension classes.

**Comment by elliott5** [comment]
Sorry @PXshadow, I can't seem to make a failing example, so I hope I am wrong.

**Comment by elliott5** [comment]
OK, if you generate code for `rnd.go`, in `Rnd.hx` (line 252) you will see there is no `@:rtti` before:
```
class _NonStructT__extension {
    public static function string(nst:NonStructT):GoString {
        return (("NonStructT-int" : GoString));
    }
}
```

**Comment by elliott5** [comment]
Latest generated code LGTM as regards `@:rtti`.

**Comment by elliott5** [comment]
Just to be clear, `_typeName_` needs to contain text describing the underlying type, not the surface type.

**Comment by elliott5** [comment]
Commit #38 `static final` is working, but unfortunately there is another problem. 
There is no abstract code to say what to do when casting `@:to` an `AnyInterface`.

Go code:
```
type NonStructT uint8

func (NonStructT) String() string {
	return "NonStructT-uint8"
}
```
Generates:
```
@:rtti class _NonStructT__extension {
    static final _typeName_ = "uint8";
    public static function string(___using:NonStructT):GoString {
        return (("NonStructT-uint8" : GoString));
    }
}
    
@:using(Rnd._NonStructT__extension) typedef NonStructT = GoUInt8;
```
I was expecting it to generate:
```
@:rtti class _NonStructT__extension {
    static final _typeName_ = "uint8"; // this is the underlyingType for reflect to find via rtti
    public static function string(___using:NonStructT):GoString {
        return (("NonStructT-uint8" : GoString));
    }
}
    
@:using(Rnd._NonStructT__extension) 
abstract NonStructT( GoUInt8) from(GoUInt8) to(GoUInt8) {
        public inline function new(?ui:GoUInt = 0) {
		this = ui;
	}
	@:to inline function __promote() // this passes the surfaceType to reflect to know where the rtti data is
		return new AnyInterface({value: this, typeName: Type.getClassName(_NonStructT__extension)});
}

```

**Comment by elliott5** [comment]
I'm afraid https://github.com/go2hx/go2hx/commit/63eea543b7d390568d4c4b96b4d31ea52371de11 generates incorrect code for the example above:
```
/**
     type NonStructT uint8
**/
@:rtti @:transitive abstract NonStructT(GoUInt8) from GoUInt8 to GoUInt8 {
    @:to  inline function __promote() return new AnyInterface({ value : this, typeName : "uint8" });
}
/**
    func (NonStructT) String() string {
    	return "NonStructT-uint8"
    }
**/
function string():GoString {
        return (("NonStructT-uint8" : GoString));
    }
```

The `class _NonStructT__extension` containing `_typeName_` for the underlying type is missing and not referenced.

And the `__promote()` method if the abstract uses the underlying type  (`uint8`) rather than the class path of the surface type (`Type.getClassName(class _NonStructT__extension)`) .

**Comment by PXshadow** [comment]
Let me know if latest generation is sufficent for your needs.

**Comment by elliott5** [comment]
After the two commits above, this go code:
```
package main

import "fmt"

func main() {
	var x NonStructT = 42
	var i interface{} = x
	fmt.Println(i)
}

type NonStructT uint8

func (nst NonStructT) String() string {
	return "NonStructT-int"
}
```

Generates the Haxe code:
```
@:rtti @:transitive @:forward abstract NonStructT(GoUInt8) from GoUInt8 to GoUInt8 {
    public var _typeName_(get, never) : String;
    inline function get__typeName_() return "uint8";
    @:to inline function __promote() return new AnyInterface({ value : this, typeName : "_NonStructT__extension" });
    public function string():GoString {
        var nst = ((this : NonStructT));
        return (("NonStructT-int" : GoString));
    }
}

private function main():Void {
    var x:NonStructT = (42 : IntegerType);
    var i:AnyInterface = x;
    stdgo.fmt.Fmt.println(i);
}
```

Which expands at runtime to the following  3 `.dump` files in the `command_line_arguments` directory...

`NonStructT.dump`:
```
@:used @:rtti @:transitive @:forward
abstract command_line_arguments.NonStructT from stdgo.GoUInt8 to stdgo.GoUInt8 {}
```
This seems like an empty declaration, I'm not sure if `@:rtti` will have any effect. 

`_Main/NonStructT_Impl_.dump`:
```
@:directlyUsed @:used
private class command_line_arguments._Main.NonStructT_Impl_ {

	public static var _typeName_(get,never):String;

	static inline function get__typeName_(this:stdgo.GoUInt8) {
		return "uint8";
	}

	@:to
	static inline function __promote(this:stdgo.GoUInt8) {
		var this = new stdgo._StdGoTypes.AnyInterfaceData(@:implicitCast cast this, "_NonStructT__extension");
		return cast this;
	}

	public static function string(this:stdgo.GoUInt8) {
		var nst = cast this;
		return cast "NonStructT-int";
	}
}
```
This is Haxe auto-generated class is interesting and potentially useful but has two issues:
- The name of the class in the `@:to` function should be "command_line_arguments._Main.NonStructT_Impl_" (not "_NonStructT__extension").  Unfortunately there is no short-cut to the go2hx transpiler generating the full Haxe path itself, as the `_Impl_` class does not exist until after the compilation process.
- There is no `@:rtti` meta-tag, so the rtti data is not generated, as demonstrated in the Haxe debug code below:
```
	var cl = Type.resolveClass("command_line_arguments._Main.NonStructT_Impl_");
	trace(cl, Type.getClassFields(cl), Rtti.hasRtti(cl));
        // trace-> Class<command_line_arguments._Main.NonStructT_Impl_>,[string,get__typeName_,__promote],false
```

`_Main/Main_Fields_.dump`:
```
@:used
private class command_line_arguments._Main.Main_Fields_ {
	@:keep
	static function main() {
		var x = @:implicitCast stdgo._StdGoTypes.IntegerType_Impl_.toUInt8(@:implicitCast stdgo._StdGoTypes.IntegerType_Impl_.fromInt(42));
		var this = new stdgo._StdGoTypes.AnyInterfaceData(@:implicitCast cast cast x, "_NonStructT__extension");
		var i = cast this;
		var this = [i];
		stdgo.fmt._Fmt.Fmt_Fields_.println(@:implicitCast cast this);
	}
}
```
Note how the `@:to` function has been expanded in-line above (with the wrong type name).

So, in summary, no this new code does not meet the needs of `reflect` runtime inspection for two reasons:

1. the pathname to the Haxe auto-generated class is wrong (but could be fixed); and
2. the Haxe auto-generated class does not have a `@:rtti` meta tag (which may not be fixable).

**Comment by PXshadow** [comment]
@elliott5 thank you for the feedback honestly the reason why I went only with the abstract approach was I thought the other approach wasn't possible, I just tested it and it actually does work, quite suprised.
```haxe
function main() {
  var x:GoInt = 10;
  x.test();
  trace(x);
}

@:using(Test.GoInt__extension)
typedef _GoInt = Int;

class GoInt__extension {
	public static function test(x:GoInt) {
    trace("this works?");
  }
}


@:forward
abstract GoInt(_GoInt) from _GoInt {

}
```

I'll work on using all three of these data types in cordination to do type aliasing now.

**Comment by elliott5** [comment]
The 3rd version is nearly working...
```
@:rtti class _NonStructT__extension {
    public function string(nst:_NonStructT):GoString {
        return (("NonStructT-int" : GoString));
    }
}
@:rtti @:using(Main._NonStructT__extension) typedef _NonStructT = GoUInt8;
/**
    NonStructT uint8
**/
@:transitive @:forward abstract NonStructT(_NonStructT) from _NonStructT to _NonStructT {
    public var _typeName_(get, never) : String;
    inline function get__typeName_() return "uint8";
    @:to inline function __promote():AnyInterface return new AnyInterface({ value : this, typeName : Type.getClassName(_NonStructT__extension) });
}
```
Good: 
- the `@:to` reference is correct, so the `reflect` package can find `_NonStructT__extension` and via rtti data see the `string()` method.

Bad: 
- because the `_typeName_` field is in an abstract, it does not exist at runtime, so `reflect` cannot determine the underlying type, and as a result marks it as "invalid".

Solution:
- move the `_typeName_` code out of the abstract and into the `__extension` class.

**Comment by elliott5** [comment]
Just tried the 4th version of the code ( e3814f82f37061692bcf21ecf9c293ab68c76f4b ). 
I can only apologise @PXshadow  that this is taking so long to resolve.

```
/*
type NonStructT uint8
func (nst NonStructT) String() string {
	return "NonStructT-int"
}
*/
@:rtti class _NonStructT__extension {
    public final _typeName_ = "NonStructT";
    public function string(nst:_NonStructT):GoString {
        return (("NonStructT-int" : GoString));
    }
}
@:rtti @:using(Main._NonStructT__extension) typedef _NonStructT = GoUInt8;
@:transitive @:forward abstract NonStructT(_NonStructT) from _NonStructT to _NonStructT {
    @:to
    inline function __promote():AnyInterface return new AnyInterface({ value : this, typeName : Type.getClassName(_NonStructT__extension) });
}
```

Sorry, I did not make two things clear about the `__extension` class:
- both `_typeName_` and any methods (e.g `String`) need to be `static`, so that the `rtti` inspection code can see them and so that the methods can be called (I don't think there is ever an instance of an `__extension` class);
- the `_typeName_` must contain the name of the Go type of the underlying data, in this case `uint8`.

Please would you add the `static` keyword to each of the field declarations and adjust the `typeName` to be either the name of  the underlying Go type, or, if the underlying Go type is a named type, the path to the `@:rtti` Haxe class implementing it.

**Comment by PXshadow** [comment]
No worries at all I'm sorry for how long it's taken me to figure out @elliott5 I should've remebered to have the methods be static for static extensions, but forgot when switching out of abstract methods.

> or, if the underlying Go type is a named type, the path to the @:rtti Haxe class implementing it.

This is a little tricky to do as it requires typeName look ups for the compiler, I will most likely need to implement this for other parts as well, but for now couldn't the name extension information be obtained by the ``@:rtti`` on typedef?

**Comment by elliott5** [comment]
Thanks for your persistence @PXshadow.

I'm writing some macro helper code (and a small example macro) to find the Go type at macro time.  That may help with the underlying named type issue you describe.

**Comment by elliott5** [comment]
Just tried 5th version of non-struct named type code, works perfectly for `reflect` on simple example above! 👍

**Comment by PXshadow** [comment]
@elliott5 Yay!!! Thanks for your continued support on it as well😄 . I did not rigorously test it with the reflection code it's very possible it will still need more tweaks and that's totally fine. The macro idea for finding Go types at compile time is very good, and I've used it before with pretty good results, let me know if you need any help on it, or integration with any of the other systems in the compiler/typer.

**Comment by elliott5** [comment]
I'm pleased with `stdgo.reflect.cast_AnyInterface(expression)` as, like the inspection of `@:rtti` on classes, it gives (almost) all the type information that the project needs at no extra programming cost.

The Go type information that it does not give is the length of an array, which I have set to `-1` for the time being.

Latest version at #49 

I tried, but failed, to replace the `@:from` in `AnyInterface` with (a version of) the `cast_AnyInterface` macro that was actually called when required.

My recommended alternative is to alter the go2hx code generation: 
- instead of `( expression:AnyInterface)`
- output `stdgo.reflect.cast_AnyInterface(expression)`

**Comment by PXshadow** [comment]
``@:rtti`` no longer applies.
